File: sdks/local-rp/php/src/Transport.php
<?php

declare(strict_types=1);

namespace LinkKeys\LocalRp;

/**
 * The TCP dial seam. Deliberately narrow: this interface only *connects a
 * byte stream* to `host:port`. TLS (certificate-pin verification against
 * DNS `fp=` records) is layered on top in {@see Tls}, not here, so a test
 * double can swap out "how do I open a socket" without also faking a TLS
 * handshake.
 *
 * Per the design doc's Wire Precision ("SDK endpoint discovery and
 * pinning"): the Rust `linkkeys-rpc-client` refuses non-public peer
 * addresses as a *server-side* SSRF guard. SDKs must not inherit that
 * refusal as a default — "connecting from a LAN box to wherever
 * `_linkkeys_apis` points is the entire point of this mode." The default
 * policy here is permissive. `AddressPolicy::PUBLIC_ONLY` is an opt-in for
 * integrators who specifically want that stricter posture; nothing in this
 * package selects it automatically.
 */
interface Transport
{
    /** @return resource */
    public function dial(string $hostPort);
}

/**
 * A marker for a {@see Transport} whose `dial()` already returns a
 * stream this SDK should treat as fully secured — {@see Rpc}'s TLS-pinning
 * wrap step is skipped for it entirely.
 *
 * This exists SOLELY so an in-process test double (an entirely fake,
 * non-TLS stream — see `tests/fixtures/FakeRpc.php`'s `FakeTransport`) can
 * exercise `Rpc`'s real CSIL-RPC framing end-to-end without a real TLS
 * handshake, without shipping a runtime-settable, process-global "disable
 * TLS pinning" switch in `src/` (a prior revision had exactly that: a
 * `public static` property any code could flip, silently affecting every
 * concurrent call in the process and easy to forget to reset). Choosing to
 * skip pinning now requires writing and explicitly injecting a whole
 * `Transport` implementation that opts into this interface — a
 * type-checked, call-site-visible decision, not a hidden global.
 *
 * No `Transport` implementation intended for production use should ever
 * implement this interface.
 */
interface OpaqueTransport extends Transport
{
}

final class AddressPolicy
{
    public const PERMISSIVE = 'permissive';
    public const PUBLIC_ONLY = 'public_only';
}

class TransportError extends \RuntimeException
{
}

final class AddressDeniedError extends TransportError
{
}

final class ConnectFailedError extends TransportError
{
}

/**
 * Default {@see Transport}: a plain blocking TCP socket via
 * `stream_socket_client`, gated only by `$policy` (permissive unless the
 * caller opts into {@see AddressPolicy::PUBLIC_ONLY}).
 */
final class StdTransport implements Transport
{
    public string $policy;
    public float $connectTimeoutSeconds;
    public float $ioTimeoutSeconds;

    public function __construct(string $policy = AddressPolicy::PERMISSIVE, float $connectTimeoutSeconds = 10.0, float $ioTimeoutSeconds = 30.0)
    {
        $this->policy = $policy;
        $this->connectTimeoutSeconds = $connectTimeoutSeconds;
        $this->ioTimeoutSeconds = $ioTimeoutSeconds;
    }

    /** @return resource */
    public function dial(string $hostPort)
    {
        $lastColon = strrpos($hostPort, ':');
        if ($lastColon === false) {
            throw new ConnectFailedError("{$hostPort}: missing port");
        }
        $host = substr($hostPort, 0, $lastColon);
        $port = substr($hostPort, $lastColon + 1);

        if ($this->policy === AddressPolicy::PUBLIC_ONLY) {
            $ip = $this->resolveOneAddress($host);
            if ($ip !== null && self::isNonPublic($ip)) {
                throw new AddressDeniedError("{$ip}: refusing non-public address under AddressPolicy::PUBLIC_ONLY");
            }
        }

        $errno = 0;
        $errstr = '';
        $stream = @stream_socket_client(
            "tcp://{$host}:{$port}",
            $errno,
            $errstr,
            $this->connectTimeoutSeconds,
            STREAM_CLIENT_CONNECT
        );
        if ($stream === false) {
            throw new ConnectFailedError("{$hostPort}: {$errstr} ({$errno})");
        }
        stream_set_timeout($stream, (int) $this->ioTimeoutSeconds);
        return $stream;
    }

    private function resolveOneAddress(string $host): ?string
    {
        if (filter_var($host, FILTER_VALIDATE_IP)) {
            return $host;
        }
        $ip = gethostbyname($host);
        return $ip !== $host ? $ip : null;
    }

    /**
     * True for loopback/private/link-local/CGNAT/documentation/unspecified
     * addresses. Only consulted under {@see AddressPolicy::PUBLIC_ONLY},
     * never by default.
     */
    public static function isNonPublic(string $ip): bool
    {
        if (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4) !== false) {
            $isPublic = filter_var(
                $ip,
                FILTER_VALIDATE_IP,
                FILTER_FLAG_IPV4 | FILTER_FLAG_NO_PRIV_RANGE | FILTER_FLAG_NO_RES_RANGE
            );
            if ($isPublic === false) {
                return true;
            }
            $octets = array_map('intval', explode('.', $ip));
            // CGNAT 100.64.0.0/10.
            if ($octets[0] === 100 && ($octets[1] & 0xC0) === 0x40) {
                return true;
            }
            return false;
        }

        // IPv6.
        $packed = @inet_pton($ip);
        if ($packed === false) {
            return true; // Unparseable: fail closed under PUBLIC_ONLY.
        }
        if ($ip === '::1' || $ip === '::') {
            return true;
        }
        $first2 = (ord($packed[0]) << 8) | ord($packed[1]);
        $linkLocal = ($first2 & 0xffc0) === 0xfe80;
        $ula = ($first2 & 0xfe00) === 0xfc00;
        $multicast = ord($packed[0]) === 0xff;
        if ($linkLocal || $ula || $multicast) {
            return true;
        }
        // IPv4-mapped IPv6 (::ffff:a.b.c.d).
        if (str_starts_with($packed, str_repeat("\x00", 10) . "\xff\xff")) {
            $v4 = inet_ntop(substr($packed, 12, 4));
            return self::isNonPublic($v4);
        }
        return false;
    }
}

File: sdks/local-rp/ocaml/lib/transport.ml
(* The TCP dial seam.

   Mirrors [sdks/local-rp/rust/src/transport.rs] / the Ruby port's
   [transport.rb]. Deliberately narrow: this seam only *connects a byte
   stream* to [host:port]. TLS (certificate-pin verification against DNS
   [fp=] records) is layered on top in [tls_client.ml], not here, so a test
   double can swap out "how do I open a socket" without also faking a TLS
   handshake.

   Per the design doc's Wire Precision ("SDK endpoint discovery and
   pinning"): the Rust [linkkeys-rpc-client]'s non-public-address refusal
   is a SERVER-SIDE SSRF guard and must NOT be inherited as this SDK's
   default -- "connecting from a LAN box to wherever [_linkkeys_apis]
   points is the entire point of this mode." The default policy here is
   [Permissive]. [Public_only] is an opt-in for integrators who
   specifically want that stricter posture; nothing in this package selects
   it automatically. *)

exception Connect_failed of string
exception Address_denied of string

type address_policy =
  | Permissive
  | Public_only

(* An injectable transport is just "how do I open a byte-stream socket to
   host:port" -- a one-function record rather than a module type, so a
   test fake is a one-line value rather than a first-class-module
   wrapper. *)
type t = { dial : string -> Unix.file_descr }

(* True for loopback/private/link-local/CGNAT/documentation/unspecified
   addresses. Only consulted under [Public_only], never by default. *)
let non_public (ip : Unix.inet_addr) : bool =
  let s = Unix.string_of_inet_addr ip in
  if String.contains s ':' then
    (* IPv6 *)
    let lower = String.lowercase_ascii s in
    lower = "::1" || lower = "::"
    || String.length lower >= 5 && String.sub lower 0 5 = "fe80:"
    || String.length lower >= 2 && String.sub lower 0 2 = "fc"
    || String.length lower >= 2 && String.sub lower 0 2 = "fd"
    || String.length lower >= 3 && String.sub lower 0 3 = "ff0"
  else begin
    let parts = String.split_on_char '.' s |> List.map int_of_string in
    match parts with
    | [ a; b; _c; _d ] ->
      let cgnat = a = 100 && b land 0xc0 = 0x40 in
      let loopback = a = 127 in
      let private_range = a = 10 || (a = 172 && b >= 16 && b <= 31) || (a = 192 && b = 168) in
      let link_local = a = 169 && b = 254 in
      let unspecified = s = "0.0.0.0" in
      let broadcast = s = "255.255.255.255" in
      let documentation =
        (a = 192 && b = 0 && (match parts with [ _; _; c; _ ] -> c = 2 | _ -> false))
        || (a = 198 && b = 51)
        || (a = 203 && b = 0)
      in
      cgnat || loopback || private_range || link_local || unspecified || broadcast || documentation
    | _ -> false
  end

(* Default Transport: a plain blocking TCP socket, gated only by [policy]
   (permissive unless the caller opts into [Public_only]). *)
module Std_transport = struct
  type config = { policy : address_policy; connect_timeout : float; io_timeout : float }

  let create ?(policy = Permissive) ?(connect_timeout = 10.0) ?(io_timeout = 30.0) () : config =
    { policy; connect_timeout; io_timeout }

  let split_host_port (host_port : string) : string * int =
    match String.rindex_opt host_port ':' with
    | None -> raise (Connect_failed (Printf.sprintf "%s: missing port" host_port))
    | Some idx ->
      let host = String.sub host_port 0 idx in
      let port_str = String.sub host_port (idx + 1) (String.length host_port - idx - 1) in
      if host = "" then raise (Connect_failed (Printf.sprintf "%s: missing host" host_port));
      (match int_of_string_opt port_str with
      | None -> raise (Connect_failed (Printf.sprintf "%s: invalid port" host_port))
      | Some port -> (host, port))

  let connect_with_timeout (sock : Unix.file_descr) (sockaddr : Unix.sockaddr) (timeout : float) : unit =
    Unix.set_nonblock sock;
    (try Unix.connect sock sockaddr with
    | Unix.Unix_error ((Unix.EINPROGRESS | Unix.EWOULDBLOCK), _, _) -> (
      match Unix.select [] [ sock ] [] timeout with
      | _, [ _ ], _ -> (
        match Unix.getsockopt_error sock with
        | None -> ()
        | Some err -> raise (Connect_failed (Unix.error_message err)))
      | _ ->
        (try Unix.close sock with _ -> ());
        raise (Connect_failed "connect timed out")));
    Unix.clear_nonblock sock

  let dial (cfg : config) (host_port : string) : Unix.file_descr =
    let host, port = split_host_port host_port in
    let addrs =
      try Unix.getaddrinfo host (string_of_int port) [ Unix.AI_SOCKTYPE Unix.SOCK_STREAM ]
      with _ -> raise (Connect_failed (Printf.sprintf "%s: resolve failed" host_port))
    in
    if addrs = [] then raise (Connect_failed (Printf.sprintf "%s: no address resolved" host_port));
    let rec try_addrs last_err = function
      | [] -> raise (match last_err with Some e -> e | None -> Connect_failed (Printf.sprintf "%s: no address resolved" host_port))
      | (ai : Unix.addr_info) :: rest -> (
        match ai.ai_addr with
        | Unix.ADDR_INET (ip, _) when cfg.policy = Public_only && non_public ip ->
          try_addrs
            (Some (Address_denied (Printf.sprintf "%s: refusing non-public address under Public_only" (Unix.string_of_inet_addr ip))))
            rest
        | _ -> (
          try
            let sock = Unix.socket ai.ai_family ai.ai_socktype ai.ai_protocol in
            connect_with_timeout sock ai.ai_addr cfg.connect_timeout;
            Unix.setsockopt_float sock Unix.SO_RCVTIMEO cfg.io_timeout;
            Unix.setsockopt_float sock Unix.SO_SNDTIMEO cfg.io_timeout;
            sock
          with
          | Connect_failed _ as e -> try_addrs (Some e) rest
          | Unix.Unix_error (err, _, _) -> try_addrs (Some (Connect_failed (Printf.sprintf "%s: %s" host_port (Unix.error_message err)))) rest))
    in
    try_addrs None addrs

  let to_transport (cfg : config) : t = { dial = dial cfg }
end

let default_transport : t = Std_transport.to_transport (Std_transport.create ())
